planning/behavior_velocity_planner/src/scene_module/speed_bump/scene.cpp

```diff
@@ -0,0 +1,187 @@
+// Copyright 2020 Tier IV, Inc., Leo Drive Teknoloji A.Ş.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "scene_module/speed_bump/scene.hpp"
+
+#include "motion_utils/motion_utils.hpp"
+#include "scene_module/crosswalk/util.hpp"
+#include "tier4_autoware_utils/tier4_autoware_utils.hpp"
+
+#include <rclcpp/rclcpp.hpp>
+
+namespace behavior_velocity_planner
+{
+using motion_utils::calcLongitudinalOffsetPoint;
+using tier4_autoware_utils::createPoint;
+
+namespace
+{
+}  // namespace
+
+SpeedBumpModule::SpeedBumpModule(
+  const int64_t module_id, const int64_t lane_id,
+  const lanelet::autoware::SpeedBump & speed_bump_reg_elem, const PlannerParam & planner_param,
+  const rclcpp::Logger & logger, const rclcpp::Clock::SharedPtr clock)
+: SceneModuleInterface(module_id, logger, clock),
+  module_id_(module_id),
+  lane_id_(lane_id),
+  speed_bump_reg_elem_(std::move(speed_bump_reg_elem)),
+  planner_param_(planner_param),
+  state_(State::SLOW_DOWN),
+  passed_slow_start_point_(false)
+{
+  // point.x : height [m] -- point.y : speed [m/s]
+  Point32 p1;
+  Point32 p2;
+
+  p1.x = 0.05;  // min speed bump height
+  p2.x = 0.30;  // max speed bump height
+
+  p1.y = 2.78;  // [10 kph] max speed for any speed bump
+  p2.y = 1.39;  // [5 kph] min speed for any speed bump
+
+  auto const & constants = getLinearEquation(p1, p2);
```

Comment:
This algorithm does not make sense to me.
Any reason to use this? (e.g. paper, experimental results, ...)
![image](https://user-images.githubusercontent.com/20228327/197661294-509246f8-031d-4d33-b4fd-8283a61f1f1f.png)

IMO, just inserting a constant velocity to the speed bump spot in the map sounds better to me.

Reply:
@takayuki5168 

Thank you for the comment,

> Any reason to use this? (e.g. paper, experimental results, ...)

Just wanted to calculate simply the velocity wrt speed bump height considering we would improve later. I don't know exactly how it is in Japan but in Turkey there are very different kinds of speed bumps with various heights and lenghts so I believe calculating the related velocity should be done in autoware.

> IMO, just inserting a constant velocity to the speed bump spot in the map sounds better to me

Sorry, It is not clear to me what you meant. Do you mean

1- We dont need this module at all and lets just change related lane's speed_limit on lanelet

or

2- Calculating the velocity considering bump height is not necessary so lets assign same velocity for all the speed bumps

Reply:
@takayuki5168 in future, we might even need to detect speed bumps dynamically in the perception layer and add slow down behavior. So it is better to have this as a separate module.

## Previous work on detecting speed bumps:
- Real Time Detection of Speed Hump/Bump and Distance Estimation with Deep Learning using GPU and ZED Stereo Camera 
  - https://www.sciencedirect.com/science/article/pii/S1877050918320295
- Speed Bump Detection System
  - https://cecas.clemson.edu/cvel/auto/AuE835_Projects_2011/Potluri_project.html
- Speed Bump Detection Using Accelerometric Features: A Genetic Algorithm Approach
  - https://www.mdpi.com/1424-8220/18/2/443
- Real-Time Speed Bump Detection Using Image Segmentation for Autonomous Vehicles
  - https://www.researchgate.net/publication/336671015_Real-Time_Speed_Bump_Detection_Using_Image_Segmentation_for_Autonomous_Vehicles
- Real time speed bump detection using Gaussian filtering and connected component approach
  - https://ieeexplore.ieee.org/document/7583981

## Previous work on velocity planning/control on speed bumps:
- Vehicle velocity optimization for ride comfort improvement of automated driving in speed hump conditions
  - https://www.sciencedirect.com/science/article/pii/S2405896321015998
- Research on adaptive speed control method of an autonomous vehicle passing a speed bump on the highway based on a genetic algorithm
  - https://ms.copernicus.org/articles/13/647/2022/
- Deceleration Planning Algorithm Based on Classified Multi-Layer Perceptron Models for Smart Regenerative Braking of EV in Diverse Deceleration Conditions
  - https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6766928/
    - "Therefore, the individual MLP models were designed according to three different deceleration conditions: car-following, **speed bump**, and intersection."

These links are to show the need to treat speed bumps as unique objects to label and act.

These can still be discussed:
- The slow down curve
  - linear/parabolic
  - acceleration (time window)
- Speed on the speed bump based on the speed bump type

Reply:
@takayuki5168 

After the this week's ASWG meeting, an optional tag is [added](https://github.com/autowarefoundation/autoware.universe/pull/647/commits/21f34e116fa0ed38591e13d84eed7b5f8be8926d) to speed_bump annotation so it would override the calculated speed. 

Also the limits of speed and height for the speed calculation is [configurable now](https://github.com/autowarefoundation/autoware.universe/pull/647/commits/d50589e6c0b3ce67379391a79527856c778c5723) via config file.

Reply:
> Sorry, It is not clear to me what you meant. Do you mean
> 1- We dont need this module at all and lets just change related lane's speed_limit on lanelet
> or
> 2- Calculating the velocity considering bump height is not necessary so lets assign same velocity for all the speed bumps

Sorry to be not clear. Neither 1 or 2.

As you suggest, I agree to add this module in behavior_velocity_planner.
What I wanted to say is, when manually creating a lanelet map with speed bumps, just setting velocity (not the same velocity to all the bumps) to each bump element on a map (= each speed bump has each pre-defined velocity on a map) sounds better to me.

But now I understand that in the future you want to dynamically detect the speed bumps and calculate target velocity based on the recognition result. And as a first step, you calculate target velocity from the speed bump element with manually pre-defined height in the map.

Still, setting velocity rather than height to each speed bump element on the map sounds a bit better to me. The reason you want to set not velocity but height to each speed bump element on the map is because height will be detected soon, right?

Reply:
@takayuki5168 

Thank you for your reply.

What sounds better to you (setting different speeds for each speed_bump on the map) is now available with [the optional tag](https://github.com/autowarefoundation/autoware_common/pull/120#:~:text=As%20an%20option,/%3E%0A%20%20%3C/way%3E) `slow_down_speed` :) 

However, whether we detect the speed bump or not, my thinking is that deciding the proper speed should be done in autoware for 2 reasons:

1) I believe HAD-Maps should hold the sementic information related to the primitive. In our case while the height of the speed bump as well as the shape of it (in 2D) is related to the primitive, the speed specified on the map for speed bump is not related. In my perspective HAD-Maps play a role as facilitator of perceiving things around; sometimes even a substitute for it. While the annotation of a traffic light is a facilitator of perceiving the color of the light, annotating the lane boundries is a substitute for perception. I think having higher levels of information than these on HAD-Maps means overloading and kind of like hard-coding. In this case while getting the height and location information belongs to the related speed bump is a substitution of perception, the speed information is unrelated.

2) In my opinion, HAD-Maps should be environment specific tools not vehicle specific since different kind of vehicles can use the same HAD-Map. Different vehicles need different speed profiles of course. Therefore a truck and a mini delivery car need different target speed for the same speed bump.

Reply:
Now I got your intention. Sounds good!